docs/posts/flink-unified-stream-batch/index.qmd

```diff
@@ -0,0 +1,414 @@
+results look promising, the features and models would then be deployed into
+production. Oftentimes, there is a code handover in this process, and a
+dedicated team of developers would be responsible for rewriting the logic for
+production, as a streaming workload.
+
+In both cases, there is a huge amount of technical overhead. If there is a
+streamlined architecture, using a unified API, much of this overhead can be
+avoided. As a platform engineer, you no longer need to worry about maintaining
+two separate architectures and codebases. As a data scientist or a machine
+learning engineer, you can write one single workload that can execute both on
+batch and streaming backends. Wouldn’t that be amazing?
+
+## Ibis unifies batch and streaming
+
+Enter Ibis. Ibis unifies batch and streaming with a single API. It decouples
+the dataframe API from backend execution, so that the logic for defining data
+transformations is unaffected by implementation discrepancies across backend
+engines. There is also an ongoing effort to further increase interoperability
+across different languages and systems via a standard query plan intermediate
+representation (IR), using a library called
+[`Substrait`](https://substrait.io/).
+
+What does this actually look like? For example, Ibis allows users to define
+window aggregations using the [`over()`
+method](https://ibis-project.org/reference/expression-tables#ibis.expr.types.groupby.GroupedTable.over).
+When executed on the Flink backend, this translates into [Flink’s over
+aggregation query](https://nightlies.apache.org/flink/flink-docs-release-1.18/docs/dev/table/sql/queries/over-agg/)
+and outputs an aggregated value for every input row over a range of ordered
+rows. On streaming data, aggregation results are continuously computed and
+written into data sinks (e.g., Kafka, Redis) as records are received at and
+consumed from the upstream data source (e.g., Kafka, Change Data Capture). In
+pandas, the conceptual analog is [windowing
+operation](https://pandas.pydata.org/docs/user_guide/window.html). Results are
+computed by looking back the length of the window from the current observation,
+but can be computed all at once because batch data is static.
+
+Another great example is deduplication. In Flink SQL, this looks something like this:
+
+```sql
+SELECT [column_list]
+FROM (
+   SELECT [column_list],
+     ROW_NUMBER() OVER ([PARTITION BY col1[, col2...]]
+       ORDER BY time_attr [asc|desc]) AS rownum
+   FROM table_name)
+WHERE rownum = 1
+```
+In a database like Postgres, this could be as simple as
+```sql
+SELECT DISTINCT t0.`string_col`, t0.`int_col`
+FROM functional_alltypes t0
+```
+And in `pandas`, you would use the method `drop_duplicates()`:
+```python
+df.drop_duplicates()
+```
+
+::: {.callout-note}
+We’re working on supporting deduplication via `distinct()` in Flink backend and
+this feature should be available before the next major release!
```

Comment:
are you sure? wouldn't publicly promise features in a given release unless we are very sure

Reply:
changing it to "soon" 😂